### DISCUSSION TITLE
typegoose graphql update subdocument

### DISCUSSION THREAD

**Discussion by Sed9yDataBank** [MichalLytek/type-graphql#959] [Q&A]
I am trying to write a mutation for updating a subdocument.
The subdocument has multiple fields.
This is how it's done with mongo:
```shell
db.getCollection('positions').findOneAndUpdate({_id: ObjectId("60f49c9961fd9384783ef912"), 
  "progressions._id": ObjectId("60f4a310b4a2919386b77391")}, 
 {"$set":{"progressions.$.progressionName":"NEWWWWWWWWWWWW"}})
```
and this is what I wrote in my resolver:
```typescript
@Mutation(() => Boolean, { nullable: false })
  async updateProgression(
    @Arg("positionId", type => ObjectIdScalar) positionId: ObjectId,
    @Arg("progressionId", type => ObjectIdScalar) progressionId: ObjectId,
    @Arg("progression", () => ProgressionInput, { nullable: true }) progressionInput: ProgressionInput) {
      const position = await PositionModel.findOneAndUpdate({
                   _id: positionId as any,
                   "progressions._id": progressionId as any
        }, progressionUpdateInput, {useFindAndModify: false});
  }
```

```
@InputType()
export class ProgressionUpdateInput implements Partial<Progression> {

  @Field({ nullable: true })
  progressionName?: string;

  @Field(() => String, { nullable: true})
  progressionDate?: Date;

  @Field({ nullable: true })
  progressionRanking?: number;

  @Field({ nullable: true })
  passingStatus?: string;

  @Field({ nullable: true })
  notes?: string;
}
```

```typescript
@ObjectType()
export class Position {
    @Field(() => ObjectIdScalar)
    @prop({ auto: true })
    readonly _id: mongoose.Schema.Types.ObjectId;

    @Field()
    @prop({ required: true, unique: true })
    public positionName!: string;

    @Field(() => [Progression])
    @prop({ type: () => Progression, default: [] })
    progressions: Progression[];
}
```

My **problem** is how can I partial update my subdocument 'Progression'.
For now graphql playground returns true, but there are no updates happening.

**Comment by carlocorradini** [comment]
Two things:
1. Post ```ProgressionInput``` class.
2. *How do I pass progressionInput in the $set argument*? Ca you explain better what you are trying to achieve. Unfortunately, I don't undestand your doubt 😪

PS: I'm not familiar with ```typegoose```.

**Reply by Sed9yDataBank** [reply]
I have updated the question above, thank you for your time!
I just want to do the same as the other question but this time I want to partial update the subdocument.

**Reply by carlocorradini** [reply]
e.g: The user wants to update only ```progressionName``` and ```notes```?

**Reply by Sed9yDataBank** [reply]
Yes

**Reply by Sed9yDataBank** [reply]
The user can update fields of progression as he likes, sometimes he may update one filed and other times he may just update one field.

**Reply by carlocorradini** [reply]
The ```progressionInput``` variable in the mutation will have only this two fields (the others are left out unless you set a default value). e.g. ```{ progressionName: 'hello', notes: 'world' }```.

Update only the fields that are present in ```progressionInput```. An example using [TypeORM](https://github.com/typeorm/typeorm) (only to understand the concept):
```typescript
@Mutation(() => Boolean, { nullable: false })
async updateProgression(
    @Arg("positionId", type => ObjectIdScalar) positionId: ObjectId,
    @Arg("progressionId", type => ObjectIdScalar) progressionId: ObjectId,
    @Arg("progression", () => ProgressionInput) progressionInput: ProgressionInput) {
    await progressionRepository.update(positionId, {
      ...(progressionInput.progressionName && { progressionName: progressionInput.progressionName }),
      ...(progressionInput.progressionDate && { progressionDate: progressionInput.progressionDate }),
      ...(progressionInput.notes && { notes: progressionInput.notes}),
     /* ... More fields ... */
    });
  }
```

**Update only those fields that are present in ```progressionInput```**

**Reply by Sed9yDataBank** [reply]
So there isn't similar approach like this one? https://github.com/MichalLytek/type-graphql/discussions/962#discussioncomment-1051672
 
_The above works for a document but it didn't work for subdocument aka progression when I tried it._

**Reply by carlocorradini** [reply]
I don't understand sorry :(

**Reply by carlocorradini** [reply]
Can you make an example or describe more in details what you want to achieve

**Reply by Sed9yDataBank** [reply]
It's fine thank you anyways.

**Reply by carlocorradini** [reply]
Seriously, if I better understand I can help you 😁

**Reply by Sed9yDataBank** [reply]
Well, I understand that my problem is related to mongodb and nosql. If this was done with postgres for example, it would have been easy.
Hopefully someone with typegoose experience can answer this. 
Again thank you :)
